File: javascript/dp/1646. Get Maximum in Generated Array.js
/**
 * @param {number} n
 * @return {number}
 */
var getMaximumGenerated = function(n) {
    if (n == 0) return 0;

    let nums = [0, 1];
    for (let i=2; i<n+1; i++) {
        if (i % 2 == 0) nums.push(nums[Math.round(i/2)])
        else nums.push(nums[Math.round((i-1) / 2)] + nums[Math.round((i-1) / 2 + 1)])
    }
    return nums.reduce((a, b) => Math.max(a, b), -Infinity);  // Returns -Infinity if no parameters are provided.
};

var getMaximumGenerated = function(n) {
    if (n == 0) return 0;

    const nums = [0, 1];
    let res = 1;
    for (let i=2; i<n+1; i++) {
        if (i % 2 == 0) nums.push(nums[Math.round(i/2)])
        else nums.push(nums[Math.round((i-1) / 2)] + nums[Math.round((i-1) / 2 + 1)])

        res = Math.max(res, nums[i]);
    }
    return res; 
};

File: python/dp/1646. Get Maximum in Generated Array.py
'''
(own)
Time: O(n)
Space: O(n)
'''

class Solution:
    def getMaximumGenerated(self, n: int) -> int:
        if n == 0: return 0
        nums = [0, 1]

        for i in range(2, n+1):
            if i % 2 == 0:
                nums.append(nums[int(i/2)])
            else:
                nums.append(nums[int((i-1) / 2)] + nums[int((i-1) / 2 + 1)]) 

        return max(nums)